### PR TITLE
port overlay-lists

### PR DESCRIPTION
**Pull request by brotzeit**

### CODE REVIEW

**Comment by shaleh** [comment]
Triggering a build

File: rust_src/src/buffers.rs [hunk 1]
```diff
@@ -159,6 +159,26 @@ impl LispBufferRef {
             self.fetch_byte(n) as c_int
         }
     }
+
+    #[inline]
+    pub fn overlays_before(&self) -> Option<LispOverlayRef> {
+        let ov = self.overlays_before;
+        if ov.is_null() {
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: This looks like a pattern that might be worthwhile to pull out into its own function.

File: rust_src/src/buffers.rs [hunk 2]
```diff
@@ -359,3 +388,40 @@ fn set_buffer(buffer_or_name: LispObject) -> LispObject {
...
+fn overlay_lists() -> LispObject {
+    let cur_buf = ThreadState::current_buffer();
+    fn list_overlays(overlay: Option<LispOverlayRef>) -> LispObject {
+        let mut ol = overlay;
+        let mut ol_list = LispObject::constant_nil();
+        let mut tmp: LispObject;
+        loop {
+            match ol {
+                Some(o) => {
+                    tmp = unsafe {
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Why make this a LispObject to just unwrap it immediately? Same question for the result of `Fcons`.

File: rust_src/src/buffers.rs (hunk 1 above)
Comment (brotzeit) [reply rust_src/src/buffers.rs]: `overlays_before` and `overlays_after` seem to be used in several places in the c code. Do you think it's enough to justify a method ?

Comment (shaleh) [reply rust_src/src/buffers.rs]: I am thinking that "is_null ? None : Some(transmute)" is probably a pattern.

Comment (shaleh) [reply rust_src/src/buffers.rs]: As it is that block is repeated 3 times in this file.

Comment (birkenfeld) [reply rust_src/src/buffers.rs]: `ptr.as_ref().map(mem::transmute)` should do it

File: rust_src/src/buffers.rs [hunk 3]
```diff
@@ -359,3 +388,40 @@ fn set_buffer(buffer_or_name: LispObject) -> LispObject {
...
+/// The lists you get are copies, so that changing them has no effect.
+/// However, the overlays you get are the real objects that the buffer uses.
+#[lisp_fn]
+fn overlay_lists() -> LispObject {
+    let cur_buf = ThreadState::current_buffer();
+    fn list_overlays(overlay: Option<LispOverlayRef>) -> LispObject {
+        let mut ol = overlay;
+        let mut ol_list = LispObject::constant_nil();
+        let mut tmp: LispObject;
+        loop {
```
Comment (birkenfeld) [inline rust_src/src/buffers.rs]: If this loop over overlays occurs more often, it might be useful to define an iterator for it.

Comment (shaleh) [reply rust_src/src/buffers.rs]: 👍

**Comment by shaleh** [comment]
Toggle build

**Comment by shaleh** [comment]
trigger a build

File: rust_src/src/buffers.rs [hunk 4]
```diff
@@ -209,6 +210,18 @@ impl LispBufferRef {
             c_int::from(self.fetch_byte(n))
         }
     }
+
+    #[inline]
+    pub fn overlays_before(&self) -> Option<LispOverlayRef> {
+        let ov = unsafe { bget_overlays_before(self.as_ptr()) };
+        unsafe { ov.as_ref().map(|o| mem::transmute(o)) }
+    }
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Perhaps these two function bodies could be refactored into a `new_from_ptr` method on `LispOverlayRef`?

    pub fn overlays_before(&self) -> Option<LispOverlayRef> {
        LispOverlayRef::new_from_ptr(unsafe { bget_overlays_before(self.as_ptr()) })
    }

Maybe?

File: rust_src/src/buffers.rs [hunk 5]
```diff
@@ -225,6 +238,27 @@ impl LispOverlayRef {
...
+pub struct LispOverlayIter {
+    current: Option<LispOverlayRef>,
+}
+
+impl Iterator for LispOverlayIter {
+    type Item = LispOverlayRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.current.map(|ov| {
+            self.current = unsafe { ov.next.as_ref().map(|o| mem::transmute(o)) };
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: And here again. `new_from_ptr`.

Also, I think this method would be more idiomatically written with a `match`. It feels odd to use `map` for side-effects.

File: rust_src/src/buffers.rs [hunk 6]
```diff
@@ -434,4 +468,36 @@ pub extern "C" fn nsberror(spec: Lisp_Object) -> ! {
...
+/// The lists you get are copies, so that changing them has no effect.
+/// However, the overlays you get are the real objects that the buffer uses.
+#[lisp_fn]
+pub fn overlay_lists() -> LispObject {
+    let list_overlays = |ol: Option<LispOverlayRef>| -> LispObject {
+        let mut ol_list = LispObject::constant_nil();
+
+        if let Some(o) = ol {
+            let mut it = o.iter();
+
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: The upcoming transmute call should really be a method on `LispOverlayRef`. This is basically `as_lisp_obj` right?

File: rust_src/src/buffers.rs (hunk 4 above)
Comment (brotzeit) [reply rust_src/src/buffers.rs]: Would `from_ptr` also be ok ?

Comment (shaleh) [reply rust_src/src/buffers.rs]: Sounds reasonable.

File: rust_src/src/buffers.rs (hunk 6 above)
Comment (brotzeit) [reply rust_src/src/buffers.rs]: I don't understand which part you mean. `iter` returns a `LispOverlayIter`.

Comment (shaleh) [reply rust_src/src/buffers.rs]: My comment is intended for the while loop. Which is why I said "upcoming transmute call". If I had put the comment on the transmute line the while loop would not be legible.

File: rust_src/src/buffers.rs [hunk 7]
```diff
@@ -434,4 +474,36 @@ pub extern "C" fn nsberror(spec: Lisp_Object) -> ! {
...
+            while let Some(n) = it.next() {
+                let tmp: LispObject =
+                    unsafe { mem::transmute(LispObject::tag_ptr(n, Lisp_Type::Lisp_Misc)) };
+                ol_list = unsafe { LispObject::from_raw(Fcons(tmp.to_raw(), ol_list.to_raw())) };
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Since everything is a `LispObject` wouldn't `LispObject::cons` work better here?

Comment (brotzeit) [reply rust_src/src/buffers.rs]: I have to cast it to `Lisp_Object` for the next call of `Fcons`. Should I add a method `to_raw` for `Lisp_Cons` ?

File: rust_src/src/buffers.rs [hunk 8]
```diff
@@ -434,4 +474,36 @@ pub extern "C" fn nsberror(spec: Lisp_Object) -> ! {
...
+            let mut it = o.iter();
+
+            while let Some(n) = it.next() {
+                ol_list = unsafe {
+                    LispObject::from_raw(Fcons(n.as_lisp_obj().to_raw(), ol_list.to_raw()))
+                };
+            }
+        }
+
+        ol_list
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: You can use `as_lisp_obj` here

File: rust_src/src/buffers.rs [hunk 9]
```diff
@@ -434,4 +474,34 @@ pub extern "C" fn nsberror(spec: Lisp_Object) -> ! {
...
+            let mut it = o.iter();
+
+            while let Some(n) = it.next() {
+                ol_list = unsafe { Fcons(n.as_lisp_obj().to_raw(), ol_list) };
+            }
+        }
+
+        LispObject::from_raw(ol_list)
+    };
+
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: This closure could also be written:

    let list_overlays = |ol: Option<LispOverlayRef>| -> LispObject {
        if let Some(o) = ol { // or use match...
            let ol_list = o.iter().fold(Qnil, |(accum, n)| unsafe { Fcons(n.as_lisp_obj().to_raw(), accum) });
            LispObject::from_raw(ol_list)
        } else {
            LispObject::constant_nil()
        }
    }

Comment (brotzeit) [reply rust_src/src/buffers.rs]: Nice

Comment (brotzeit) [reply rust_src/src/buffers.rs]:
```
error[E0619]: the type of this value must be known in this context
   --> /home/fredbert/tmp/git/remacs/rust_src/src/buffers.rs:490:23
    |
490 |                 Fcons(n.as_lisp_obj().to_raw(), accum)
    |                       ^^^^^^^^^^^^^^^
```

Hmmm. I'm not sure how to solve this elegantly.

Comment (shaleh) [reply rust_src/src/buffers.rs]: My mistake, I slipped into Haskell style :-)

The lambda does not take a tuple. It should be `|accum, n|` not `|(accum, n)|`.

Comment (shaleh) [reply rust_src/src/buffers.rs]: pub fn overlay_lists() -> LispObject {
        let list_overlays = |ol: LispOverlayRef| -> LispObject {
            let ol_list = ol.iter().fold(Qnil, |accum, n| unsafe {
                Fcons(n.as_lisp_obj().to_raw(), accum)
            });
            LispObject::from_raw(ol_list)
        };
    
        let cur_buf = ThreadState::current_buffer();
        let before = cur_buf
            .overlays_before()
            .map_or_else(|| LispObject::constant_nil(), &list_overlays);
        let after = cur_buf
            .overlays_after()
            .map_or_else(|| LispObject::constant_nil(), &list_overlays);
        unsafe { LispObject::from_raw(Fcons(Fnreverse(before.to_raw()), Fnreverse(after.to_raw()))) }
    }

**Comment by shaleh** [comment]
Windows oddness. Let's try again.

**Comment by shaleh** [comment]
Ok, Windows build issue resolved. Let's try this again.

**Pull request merged**